### ISSUE TITLE
Need to document the correct behaviors user should expect from Addon Manager v6.0.

### ISSUE THREAD

**Issue by MrHohn** [kubernetes/kubernetes#36411]
The new Addon Manager being bumped up in #36008 has some fundamental behavioral changes comparing to the old one. This [README](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/README.md) explains how the new Addon Manager work and makes below claim:
> The add-ons are visible through the API (they can be listed using kubectl), but direct manipulation of these objects through Apiserver is discouraged because the system will bring them back to the original state.

However, the underlying consequences are not well covered in the case that users don't have access to the master node. Issues we've encountered listed below:
- Some applications may be broke by this new behavior, like #36194. The fact is that if the application depends on changing the addon configuration through the apiserver in order to update it, very likely this application will no longer work.
- A few addons that need to stay adjustable while running are also affected. Under this new behavior, they need to leave specific fields unset in the manifest files in order to make them tunable. Like [`kube-dns`](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns) leaves `replicas` unset for horizontal scaling and [`heapster`](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/cluster-monitoring/influxdb) leaves `resource` unset for vertical scaling.
- Those fileds left unset in the manifest files are tunable based on one precondition that the modification will not save the configuration of object in its annotation(`last-applied-configuration`). Otherwise the Addon Manager will still reconcile it due to the [three-way-merge](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L1165-L1247) strategy. For instance, modifications made by k8s go clients will work(like [horizontal autoscaler](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler) and [vertical autoscaler](https://github.com/kubernetes/contrib/tree/master/addon-resizer)). But `kubectl apply`, `kubectl edit`, `kubectl create` etc. will not work because they will update the annotation.

In short, the behavior for modifying the adjustable fields of addons are chaos and definitely not recommended to users. We should really emphasize this in docs to reduce the potential of breaking more unexpected applications.

@mikedanese 

cc @bowei @bprashanth

**Comment by bprashanth** [comment]
On Mon, Nov 7, 2016 at 6:59 PM, Zihong Zheng <notifications@github.com>
wrote:

> The new Addon Manager being bumped up in #36008
> <https://github.com/kubernetes/kubernetes/pull/36008> has some
> fundamental behavioral changes comparing to the old one. This README
> <https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/README.md>
> explains how the new Addon Manager work and makes below claim:
>
> The add-ons are visible through the API (they can be listed using
> kubectl), but direct manipulation of these objects through Apiserver is
> discouraged because the system will bring them back to the original state.
>
> However, the underlying consequences are not well covered in the case that
> users don't have access to the master node. Issues we've encountered listed
> below:
>
>    - Some applications may be broke by this new behavior, like #36194
>    <https://github.com/kubernetes/kubernetes/issues/36194>. The fact is
>    that if the application depends on changing the addon configuration through
>    the apiserver in order to update it, very likely this application will no
>    longer work.
>    - A few addons that need to stay adjustable while running are also
>    affected. Under this new behavior, they need to leave specific fields unset
>    in the manifest files in order to make them tunable. Like kube-dns
>    <https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns>
>    leaves replicas unset for horizontal scaling and heapster
>    <https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/cluster-monitoring/influxdb>
>    leave s resource unset for vertical scaling.
>    - Those filed left unset in the manifest files are tunable based on
>    one precondition that the modification will not save the configuration of
>    object in its annotation(last-applied-configuration). Otherwise the
>    Addon Manager will still reconcile it due to the three-way-merge
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L1165-L1247>
>    strategy. For instance, modifications made by k8s go clients will work(like horizontal
>    autoscaler
>    <https://github.com/kubernetes-incubator/cluster-proportional-autoscaler>
>    and vertical autoscaler
>    <https://github.com/kubernetes/contrib/tree/master/addon-resizer>).
>    But kubectl apply, kubectl edit, kubectl create etc. will not work
>    because they will update the annotation.
>
> kubectl patch?


>
>
> In short, the behavior for modifying the adjustable fields of addons are
> chaos and definitely not recommended to users. We should really emphasize
> this in docs to reduce the potential of breaking more unexpected
> applications.
>

maybe we should write an admission controller to reject updates to addons
on the whole?

**Comment by MrHohn** [comment]
> kubectl patch?

I think `kubectl patch` will work. And `kubectl scale` also work.

> maybe we should write an admission controller to reject updates to addons
on the whole?

But some of the addons do need updates, like `replicas` for kube-dns and `resource` for heapster. We might want users to be able to scale `kube-dns` themselves if dns autoscaling is not turned on.
